task-queue/src/main/java/org/hawkular/metrics/tasks/impl/TaskSchedulerImpl.java

```diff
@@ -0,0 +1,626 @@
+            delegate.onCompleted();
+        }
+
+        @Override
+        public void onError(Throwable e) {
+            delegate.onError(e);
+        }
+
+        @Override
+        public void onNext(Task2 task2) {
+            try {
+                delegate.onNext(task2);
+            } catch (Exception e) {
+                logger.warn("Execution of {} failed", task2);
+            }
+        }
+    }
+
+    /**
+     * Subscribe a callback that will be responsible for executing tasks.
+     *
+     * @param onNext The task execution callback
+     * @return A subscription which can be used to stop receiving task notifications.
+     */
+    public Subscription subscribe(Action1<Task2> onNext) {
+        return taskSubject.subscribe(onNext);
+    }
+
+    /**
+     * Subscribe a callback that will be responsible for executing tasks.
+     *
+     * @param subscriber The callback
+     * @return A subscription which can be used to stop receiving task notifications.
+     */
+    public Subscription subscribe(Subscriber<Task2> subscriber) {
+        return taskSubject.subscribe(new SubscriberWrapper(subscriber));
+    }
+
+    public Observable<Long> getFinishedTimeSlices() {
+        return tickSubject;
+    }
+
+    /**
+     * Starts the scheduler so that it starts emitting tasks for execution.
+     *
+     * @return An observable that emits leases that are processed by this TaskScheduler
+     * object. The observable emits a lease only after it has been fully processed. This
+     * means all tasks belonging to the task have been executed, and the lease has been
+     * marked finished. Note that the observable is hot; in other words, it emits leases
+     * regardless of wehther or not there are any subscriptions.
+     */
+    @Override
+    public Observable<Lease> start() {
+        Observable<Date> seconds = createTicks();
+        Observable<Lease> leases = seconds.flatMap(this::getAvailableLeases);
+        Observable<Lease> processedLeases = Observable.create(subscriber -> {
+            leasesSubscription = leases.subscribe(
+                    lease -> {
+                        logger.debug("Loading tasks for {}", lease);
+                        CountDownLatch latch = new CountDownLatch(1);
+                        getQueue(lease)
+                                .observeOn(tasksScheduler)
+                                .groupBy(Task2Impl::getGroupKey)
+                                .flatMap(group -> group.flatMap(this::execute).map(this::rescheduleTask))
+                                .subscribe(
+                                        task -> logger.debug("Finished executing {}", task),
+                                        t -> logger.warn("There was an error observing tasks", t),
+                                        () -> {
+                                            Date timeSlice = new Date(lease.getTimeSlice());
+                                            // TODO We need error handling here
```

Comment:
> Well, it wasn't about that blocking behavior, but instead of error handling

Gotcha. Robust error handling is definitely needed. ```executeMultipleTasksFromMultipleGroupsInDifferentQueues``` in TaskSchedulerITest does verify that tasks execute in the correct order, but only for a single time slice. I can add another test that verifies execution order for multiple time slices. I can also include long running tasks that cause scheduling delays. 

>Because this isn't in use yet...

This is in use for counter rates. The counter rates REST integration test is disabled in this PR. The work I have done for [HWKMETRICS-199](https://issues.jboss.org/browse/HWKMETRICS-199) in the [hwkmetrics-199 branch](https://github.com/hawkular/hawkular-metrics/tree/hwkmetrics-199) re-enables the test with support for the virtual clock in REST API tests.